File: php/tests/Migrations/MigrationRunnerTest.php
<?php

declare(strict_types=1);

namespace SimpleOrm\Tests\Migrations;

use PDO;
use PHPUnit\Framework\Attributes\Test;
use PHPUnit\Framework\TestCase;
use SimpleOrm\Dialect\SqliteDialect;
use SimpleOrm\Errors\SimpleOrmException;
use SimpleOrm\Metadata\EntityMapLoader;
use SimpleOrm\Migrations\MigrationEntry;
use SimpleOrm\Migrations\MigrationRunner;
use SimpleOrm\Migrations\MigrationSet;
use SimpleOrm\Migrations\MigrationSql;
use SimpleOrm\Migrations\MigrationState;
use SimpleOrm\Migrations\MigrationStep;
use SimpleOrm\Migrations\MigrationVersion;
use SimpleOrm\Migrations\SqlVersion;
use SimpleOrm\Migrations\SqlVersionStep;
use SimpleOrm\Migrations\TableActions;
use SimpleOrm\Migrations\TableMigration;
use SimpleOrm\Migrations\VersionBuilder;
use SimpleOrm\Tests\Migrations\Fixtures\Runner\RunnerGadget;
use SimpleOrm\Tests\Support\TempDatabase;

/**
 * Milestone 5 (PHP port): the versioned migration runner, per scenario and per
 * code. Mirrors `dotnet/tests/SimpleOrm.Tests/MigrationRunnerTests.cs`.
 * Structural composition violations (`MIG-001`-`MIG-004`) are `MigrationSet`'s
 * own area, covered by `MigrationSetTest`; this file exercises what only the
 * runner can: applying, reverting, checksums against recorded history, atomic
 * failure, and baseline.
 */
final class MigrationRunnerTest extends TestCase
{
    private TempDatabase $database;

    private SqliteDialect $dialect;

    private PDO $connection;

    private EntityMapLoader $maps;

    protected function setUp(): void
    {
        $this->database = TempDatabase::create();
        $this->dialect = new SqliteDialect();
        $this->connection = $this->dialect->createConnection($this->database->connectionString());
        $this->maps = new EntityMapLoader();
    }

    protected function tearDown(): void
    {
        unset($this->connection);
        $this->database->delete();
    }

    private function runner(SqlVersion ...$versions): MigrationRunner
    {
        return new MigrationRunner($this->connection, $this->dialect, $this->maps, MigrationSet::of(...$versions));
    }

    private static function widgets1(string $sql = 'create table widgets (id INTEGER PRIMARY KEY, name TEXT) STRICT'): SqlVersion
    {
        return new SqlVersion(1, new SqlVersionStep('widgets', 'create', [$sql], ['drop table widgets']));
    }

    #[Test]
    public function migrate_applies_pending_versions_and_is_idempotent(): void
    {
        $runner = $this->runner(self::widgets1());

        self::assertTrue($runner->hasPending());
        self::assertSame(1, $runner->migrate());
        self::assertSame(0, $runner->migrate());
        self::assertFalse($runner->hasPending());

        $status = $runner->status();
        self::assertCount(1, $status);
        self::assertSame(MigrationState::Applied, $status[0]->state);
    }

    #[Test]
    public function actions_reorder_to_rename_add_remove_with_hooks_in_place(): void
    {
        $gadgetV1 = new SqlVersion(1, new SqlVersionStep(
            'runner_gadgets',
            'create',
            [
                'create table runner_gadgets (id INTEGER PRIMARY KEY, label TEXT, legacy TEXT) STRICT',
                "insert into runner_gadgets (label, legacy) values ('hello', 'old')",
            ],
            ['drop table runner_gadgets'],
        ));

        $step = new class extends TableMigration {
            public function version(): int
            {
                return 2;
            }

            public function description(): string
            {
                return 'Restructure';
            }

            public function entityClass(): string
            {
                return RunnerGadget::class;
            }

            // Deliberately declared out of order: the renderer must run rename -> add -> remove.
            public function action(TableActions $actions): void
            {
                $actions->addColumn('label', 'TEXT')->post("update runner_gadgets set label = 'L-' || title");
                $actions->removeColumn('legacy')->pre('update runner_gadgets set title = title || \'-\' || legacy');
                $actions->renameColumn('label', 'title');
            }

            public function down(TableActions $actions): void
            {
                $actions->removeColumn('label');
            }

            public function preDown(MigrationSql $sql): void
            {
                $sql->sql('update runner_gadgets set title = label'); // works only while label still exists
            }

            public function postDown(MigrationSql $sql): void
            {
                $sql->sql("update runner_gadgets set title = title || '!'"); // runs after the DDL
            }
        };
        $gadgetV2 = new class($step) extends MigrationVersion {
            public function __construct(private readonly MigrationStep $step)
            {
            }

            public function version(): int
            {
                return 2;
            }

            public function compose(VersionBuilder $version): void
            {
                $version->apply($this->step);
            }
        };

        $set = MigrationSet::of($gadgetV1, $gadgetV2);
        $runner = new MigrationRunner($this->connection, $this->dialect, $this->maps, $set);
        $runner->migrate();

        $read = $this->connection->prepare('select title, label from runner_gadgets');
        $read->execute();
        $row = $read->fetch(PDO::FETCH_ASSOC);
        self::assertSame('hello-old', $row['title']); // pre-remove ran after the add group, before remove
        self::assertSame('L-hello', $row['label']); // post-add backfilled from the renamed column

        // Down: PreDown (needs label) -> DDL (drops label) -> PostDown.
        $runner->migrateDown(1);
        $titleQuery = $this->connection->query('select title from runner_gadgets');
        self::assertSame('L-hello!', $titleQuery->fetchColumn());
    }

    #[Test]
    public function down_reverts_in_reverse_and_requires_down_statements(): void
    {
        $reversible = $this->runner(self::widgets1());
        $reversible->migrate();
        self::assertSame(1, $reversible->migrateDown(0));
        $status = $reversible->status();
        self::assertFalse(self::anyState($status, MigrationState::Applied));

        $irreversible = $this->runner(new SqlVersion(
            1,
            new SqlVersionStep('widgets', 'create', ['create table widgets (id INTEGER PRIMARY KEY) STRICT']),
        ));
        $irreversible->migrate();

        try {
            $irreversible->migrateDown(0);
            self::fail('expected MIG-020');
        } catch (SimpleOrmException $exception) {
            self::assertSame('MIG-020', $exception->errorCode);
        }
    }

    #[Test]
    public function checksum_drift_and_unknown_history_fail_before_executing(): void
    {
        $this->runner(self::widgets1())->migrate();

        $drifted = $this->runner(self::widgets1('create table widgets (id INTEGER PRIMARY KEY, extra TEXT) STRICT'));
        try {
            $drifted->migrate();
            self::fail('expected MIG-010');
        } catch (SimpleOrmException $exception) {
            self::assertSame('MIG-010', $exception->errorCode);
        }

        $emptied = $this->runner();
        try {
            $emptied->migrate();
            self::fail('expected MIG-011');
        } catch (SimpleOrmException $exception) {
            self::assertSame('MIG-011', $exception->errorCode);
        }
    }

    #[Test]
    public function failed_run_rolls_back_every_version(): void
    {
        $runner = $this->runner(
            self::widgets1(),
            new SqlVersion(2, new SqlVersionStep('widgets', 'boom', ['this is not sql'])),
        );

        try {
            $runner->migrate();
            self::fail('expected the bad statement to throw');
        } catch (SimpleOrmException) {
            // expected: MIG-021, but what matters here is that nothing survived.
        }

        // BEGIN IMMEDIATE run: nothing survives — not even V1.
        $clean = $this->runner(self::widgets1());
        $status = $clean->status();
        self::assertCount(1, $status);
        self::assertSame(MigrationState::Pending, $status[0]->state);
    }

    #[Test]
    public function baseline_records_without_running(): void
    {
        $runner = $this->runner(self::widgets1());
        $runner->baseline(1);

        self::assertFalse($runner->hasPending());
        self::assertSame(0, $runner->migrate());

        // The table was never actually created — baseline only records.
        $count = $this->connection
            ->query("select count(name) from sqlite_master where type = 'table' and name = 'widgets'")
            ->fetchColumn();
        self::assertSame(0, (int) $count);
    }

    /** @param list<MigrationEntry> $entries */
    private static function anyState(array $entries, MigrationState $state): bool
    {
        foreach ($entries as $entry) {
            if ($entry->state === $state) {
                return true;
            }
        }

        return false;
    }
}

File: go/orm/migrations/migration_runner_test.go
package migrations_test

import (
	"context"
	"database/sql"
	"reflect"
	"testing"

	"github.com/kosmassp/SimpleOrm/go/orm/internal/core"
	"github.com/kosmassp/SimpleOrm/go/orm/internal/metadata"
	"github.com/kosmassp/SimpleOrm/go/orm/migrations"
	samplemigrations "github.com/kosmassp/SimpleOrm/go/orm/sample/migrations"
	"github.com/kosmassp/SimpleOrm/go/orm/sqlite"
)

// Milestone-equivalent (§7.23): the versioned migration runner, per scenario
// and per code. Mirrors dotnet/tests/SimpleOrm.Tests/MigrationRunnerTests.cs.

func TestRunner_MigrateAppliesOnceAndIsIdempotent(t *testing.T) {
	ctx := context.Background()
	pool := openMigrationsPool(t)
	maps := metadata.NewLoader(nil)

	set, err := migrations.NewSet(rawTableVersion(
		1001, "widgets", "create", []string{"create table widgets (id INTEGER PRIMARY KEY, name TEXT) STRICT"}, nil))
	if err != nil {
		t.Fatal(err)
	}
	runner := migrations.NewRunner(pool, sqlite.New(), maps, set, nil)

	pending, err := runner.HasPending(ctx)
	if err != nil {
		t.Fatal(err)
	}
	if !pending {
		t.Fatal("expected a pending version before migrating")
	}

	applied, err := runner.Migrate(ctx, migrations.RunOptions{})
	if err != nil {
		t.Fatal(err)
	}
	if applied != 1 {
		t.Fatalf("expected 1 applied version, got %d", applied)
	}

	again, err := runner.Migrate(ctx, migrations.RunOptions{})
	if err != nil {
		t.Fatal(err)
	}
	if again != 0 {
		t.Fatalf("re-running should be idempotent, applied %d", again)
	}

	pending, err = runner.HasPending(ctx)
	if err != nil {
		t.Fatal(err)
	}
	if pending {
		t.Fatal("expected no pending versions after migrating")
	}
}

// Gadget is a test-local fixture entity for the action-reordering test:
// TableActions must render rename -> add -> remove regardless of declaration
// order, with each action's Pre/Post hooks running in place.
type Gadget struct {
	ID    int64
	Title string
	Label string
}

func gadgetMap() *core.EntityMap {
	t := reflect.TypeFor[Gadget]()
	return core.NewEntityMap(t, core.RelationTable, "gadgets", "", "", nil, []*core.PropertyMap{
		property(t, "ID", "id", core.TypeInt64, key, generated),
		property(t, "Title", "title", core.TypeString),
		property(t, "Label", "label", core.TypeString),
	}, core.KeyDatabaseGenerated, nil, nil)
}

func gadgetLoader() *metadata.Loader {
	options := (&metadata.Options{}).Register(explicitMap{entityType: reflect.TypeFor[Gadget](), m: gadgetMap()})
	return metadata.NewLoader(options)
}

// V1102_Restructure is deliberately declared out of execution order: the
// renderer must run rename -> add -> remove, and PreDown/Down/PostDown are
// the manual rollback override.
type V1102_Restructure struct {
	migrations.TableMigration[Gadget]
}

func (V1102_Restructure) Action(a *migrations.TableActions) {
	a.AddColumn("label", "TEXT").Post("update gadgets set label = 'L-' || title")
	a.RemoveColumn("legacy").Pre("update gadgets set title = title || '-' || legacy")
	a.RenameColumn("label", "title")
}

func (V1102_Restructure) Down(a *migrations.TableActions) { a.RemoveColumn("label") }

func (V1102_Restructure) PreDown(sql *migrations.MigrationSQL) {
	sql.SQL("update gadgets set title = label")
}

func (V1102_Restructure) PostDown(sql *migrations.MigrationSQL) {
	sql.SQL("update gadgets set title = title || '!'")
}

type V1102 struct{}

func (V1102) Compose(v *migrations.VersionBuilder) { v.Apply(V1102_Restructure{}) }

func TestRunner_ActionsReorderAndHooksRunInPlace(t *testing.T) {
	ctx := context.Background()
	pool := openMigrationsPool(t)

	gadgetV1 := rawTableVersion(1101, "gadgets", "create", []string{
		"create table gadgets (id INTEGER PRIMARY KEY, label TEXT, legacy TEXT) STRICT",
		"insert into gadgets (label, legacy) values ('hello', 'old')",
	}, nil)

	set, err := migrations.NewSet(gadgetV1, V1102{})
	if err != nil {
		t.Fatal(err)
	}
	runner := migrations.NewRunner(pool, sqlite.New(), gadgetLoader(), set, nil)
	if _, err := runner.Migrate(ctx, migrations.RunOptions{}); err != nil {
		t.Fatal(err)
	}

	var title, label string
	if err := pool.QueryRowContext(ctx, "select title, label from gadgets").Scan(&title, &label); err != nil {
		t.Fatal(err)
	}
	if title != "hello-old" {
		t.Errorf("expected the pre-remove hook to run after the add group: got title=%q", title)
	}
	if label != "L-hello" {
		t.Errorf("expected the post-add hook to backfill from the renamed column: got label=%q", label)
	}

	if _, err := runner.MigrateDown(ctx, 1101, migrations.RunOptions{}); err != nil {
		t.Fatal(err)
	}
	if err := pool.QueryRowContext(ctx, "select title from gadgets").Scan(&title); err != nil {
		t.Fatal(err)
	}
	if title != "L-hello!" {
		t.Errorf("expected PreDown -> DDL -> PostDown, got title=%q", title)
	}
}

func TestRunner_DownRevertsInReverseAndRequiresDownStatements(t *testing.T) {
	ctx := context.Background()
	pool := openMigrationsPool(t)
	maps := metadata.NewLoader(nil)

	reversibleSet, err := migrations.NewSet(rawTableVersion(
		1110, "widgets", "create", []string{"create table widgets (id INTEGER PRIMARY KEY) STRICT"}, []string{"drop table widgets"}))
	if err != nil {
		t.Fatal(err)
	}
	reversible := migrations.NewRunner(pool, sqlite.New(), maps, reversibleSet, nil)
	if _, err := reversible.Migrate(ctx, migrations.RunOptions{}); err != nil {
		t.Fatal(err)
	}
	reverted, err := reversible.MigrateDown(ctx, 0, migrations.RunOptions{})
	if err != nil {
		t.Fatal(err)
	}
	if reverted != 1 {
		t.Fatalf("expected 1 reverted version, got %d", reverted)
	}

	irreversibleSet, err := migrations.NewSet(rawTableVersion(
		1111, "widgets", "create", []string{"create table widgets (id INTEGER PRIMARY KEY) STRICT"}, nil))
	if err != nil {
		t.Fatal(err)
	}
	irreversible := migrations.NewRunner(pool, sqlite.New(), maps, irreversibleSet, nil)
	if _, err := irreversible.Migrate(ctx, migrations.RunOptions{}); err != nil {
		t.Fatal(err)
	}
	_, err = irreversible.MigrateDown(ctx, 0, migrations.RunOptions{})
	if core.CodeOf(err) != "MIG-020" {
		t.Fatalf("expected MIG-020 without down statements or snapshots, got %v", err)
	}
}

func TestRunner_ChecksumDriftIsMIG010(t *testing.T) {
	ctx := context.Background()
	pool := openMigrationsPool(t)
	maps := metadata.NewLoader(nil)

	original, err := migrations.NewSet(rawTableVersion(
		1120, "widgets", "create", []string{"create table widgets (id INTEGER PRIMARY KEY) STRICT"}, nil))
	if err != nil {
		t.Fatal(err)
	}
	if _, err := migrations.NewRunner(pool, sqlite.New(), maps, original, nil).Migrate(ctx, migrations.RunOptions{}); err != nil {
		t.Fatal(err)
	}

	drifted, err := migrations.NewSet(rawTableVersion(
		1120, "widgets", "create", []string{"create table widgets (id INTEGER PRIMARY KEY, extra TEXT) STRICT"}, nil))
	if err != nil {
		t.Fatal(err)
	}
	_, err = migrations.NewRunner(pool, sqlite.New(), maps, drifted, nil).Migrate(ctx, migrations.RunOptions{})
	if core.CodeOf(err) != "MIG-010" {
		t.Fatalf("expected MIG-010, got %v", err)
	}
}

func TestRunner_UnknownHistoryIsMIG011(t *testing.T) {
	ctx := context.Background()
	pool := openMigrationsPool(t)
	maps := metadata.NewLoader(nil)

	set, err := migrations.NewSet(rawTableVersion(
		1130, "widgets", "create", []string{"create table widgets (id INTEGER PRIMARY KEY) STRICT"}, nil))
	if err != nil {
		t.Fatal(err)
	}
	if _, err := migrations.NewRunner(pool, sqlite.New(), maps, set, nil).Migrate(ctx, migrations.RunOptions{}); err != nil {
		t.Fatal(err)
	}

	empty, err := migrations.NewSet()
	if err != nil {
		t.Fatal(err)
	}
	_, err = migrations.NewRunner(pool, sqlite.New(), maps, empty, nil).Migrate(ctx, migrations.RunOptions{})
	if core.CodeOf(err) != "MIG-011" {
		t.Fatalf("expected MIG-011, got %v", err)
	}
}

func TestRunner_FailedRunRollsBackEveryVersion(t *testing.T) {
	ctx := context.Background()
	pool := openMigrationsPool(t)
	maps := metadata.NewLoader(nil)

	set, err := migrations.NewSet(
		rawTableVersion(1140, "widgets", "create", []string{"create table widgets (id INTEGER PRIMARY KEY) STRICT"}, nil),
		rawTableVersion(1141, "widgets", "boom", []string{"this is not sql"}, nil),
	)
	if err != nil {
		t.Fatal(err)
	}
	_, err = migrations.NewRunner(pool, sqlite.New(), maps, set, nil).Migrate(ctx, migrations.RunOptions{})
	if core.CodeOf(err) != "MIG-021" {
		t.Fatalf("expected MIG-021, got %v", err)
	}

	// BEGIN IMMEDIATE run: nothing survives -- not even V1140.
	cleanSet, err := migrations.NewSet(rawTableVersion(
		1140, "widgets", "create", []string{"create table widgets (id INTEGER PRIMARY KEY) STRICT"}, nil))
	if err != nil {
		t.Fatal(err)
	}
	status, err := migrations.NewRunner(pool, sqlite.New(), maps, cleanSet, nil).Status(ctx)
	if err != nil {
		t.Fatal(err)
	}
	if len(status) != 1 || status[0].State != migrations.Pending {
		t.Fatalf("expected a single Pending entry, got %+v", status)
	}
}

func TestRunner_BaselineRecordsWithoutRunning(t *testing.T) {
	ctx := context.Background()
	pool := openMigrationsPool(t)
	maps := metadata.NewLoader(nil)

	set, err := migrations.NewSet(rawTableVersion(
		1150, "widgets", "create", []string{"create table widgets (id INTEGER PRIMARY KEY) STRICT"}, nil))
	if err != nil {
		t.Fatal(err)
	}
	runner := migrations.NewRunner(pool, sqlite.New(), maps, set, nil)
	if err := runner.Baseline(ctx, 1150); err != nil {
		t.Fatal(err)
	}

	pending, err := runner.HasPending(ctx)
	if err != nil {
		t.Fatal(err)
	}
	if pending {
		t.Fatal("expected no pending versions after baseline")
	}
	applied, err := runner.Migrate(ctx, migrations.RunOptions{})
	if err != nil {
		t.Fatal(err)
	}
	if applied != 0 {
		t.Fatalf("baseline should have recorded the version already, got %d applied", applied)
	}

	var count int64
	if err := pool.QueryRowContext(ctx,
		"select count(name) from sqlite_master where type = 'table' and name = 'widgets'").Scan(&count); err != nil {
		t.Fatal(err)
	}
	if count != 0 {
		t.Fatal("baseline must not create the table, only record it")
	}
}

func TestRunner_StatusReportsEveryState(t *testing.T) {
	ctx := context.Background()
	pool := openMigrationsPool(t)
	maps := metadata.NewLoader(nil)

	applied := rawTableVersion(1160, "widgets", "create", []string{"create table widgets (id INTEGER PRIMARY KEY) STRICT"}, nil)
	set1, err := migrations.NewSet(applied)
	if err != nil {
		t.Fatal(err)
	}
	if _, err := migrations.NewRunner(pool, sqlite.New(), maps, set1, nil).Migrate(ctx, migrations.RunOptions{}); err != nil {
		t.Fatal(err)
	}

	pendingVersion := rawTableVersion(1161, "widgets", "add_note", []string{"alter table widgets add column note TEXT"}, nil)
	set2, err := migrations.NewSet(applied, pendingVersion)
	if err != nil {
		t.Fatal(err)
	}
	status, err := migrations.NewRunner(pool, sqlite.New(), maps, set2, nil).Status(ctx)
	if err != nil {
		t.Fatal(err)
	}
	states := map[int64]migrations.MigrationState{}
	for _, e := range status {
		states[e.Version] = e.State
	}
	if states[1160] != migrations.Applied {
		t.Errorf("expected V1160 Applied, got %v", states[1160])
	}
	if states[1161] != migrations.Pending {
		t.Errorf("expected V1161 Pending, got %v", states[1161])
	}

	drifted := rawTableVersion(1160, "widgets", "create", []string{"create table widgets (id INTEGER PRIMARY KEY, extra TEXT) STRICT"}, nil)
	set3, err := migrations.NewSet(drifted)
	if err != nil {
		t.Fatal(err)
	}
	status3, err := migrations.NewRunner(pool, sqlite.New(), maps, set3, nil).Status(ctx)
	if err != nil {
		t.Fatal(err)
	}
	if len(status3) != 1 || status3[0].State != migrations.Drifted {
		t.Fatalf("expected a single Drifted entry, got %+v", status3)
	}

	unknownSet, err := migrations.NewSet()
	if err != nil {
		t.Fatal(err)
	}
	status4, err := migrations.NewRunner(pool, sqlite.New(), maps, unknownSet, nil).Status(ctx)
	if err != nil {
		t.Fatal(err)
	}
	if len(status4) != 1 || status4[0].State != migrations.Unknown || status4[0].ObjectName != "widgets" {
		t.Fatalf("expected a single Unknown entry for widgets, got %+v", status4)
	}
}

// TestRunner_SampleTreeRoundTripsThroughZero mirrors
// MigrationRunnerTests.Sample_migrations_apply_once_and_seed against the real
// sample tree, plus a full derived-rollback round trip (DerivedDownTests).
func TestRunner_SampleTreeRoundTripsThroughZero(t *testing.T) {
	ctx := context.Background()
	pool := openMigrationsPool(t)
	maps := metadata.NewLoader(nil)

	set, err := samplemigrations.Set()
	if err != nil {
		t.Fatal(err)
	}
	snapshots, err := migrations.SnapshotsFromFS(samplemigrations.Snapshots)
	if err != nil {
		t.Fatal(err)
	}
	runner := migrations.NewRunner(pool, sqlite.New(), maps, set, snapshots)

	pending, err := runner.HasPending(ctx)
	if err != nil {
		t.Fatal(err)
	}
	if !pending {
		t.Fatal("expected pending versions before migrating")
	}

	applied, err := runner.Migrate(ctx, migrations.RunOptions{})
	if err != nil {
		t.Fatal(err)
	}
	if applied != 10 {
		t.Fatalf("expected 10 applied versions (V0001..V0010), got %d", applied)
	}
	if again, err := runner.Migrate(ctx, migrations.RunOptions{}); err != nil || again != 0 {
		t.Fatalf("expected idempotent re-migrate, got %d, %v", again, err)
	}

	roleNames, err := readRoleNames(ctx, pool)
	if err != nil {
		t.Fatal(err)
	}
	if len(roleNames) != 2 || roleNames[0] != "admin" || roleNames[1] != "user" {
		t.Fatalf("expected [admin user] after the V0004 rename and V0005 seed, got %v", roleNames)
	}

	status, err := runner.Status(ctx)
	if err != nil {
		t.Fatal(err)
	}
	if len(status) != 16 {
		t.Fatalf("expected one row per (version, object) = 16, got %d: %+v", len(status), status)
	}
	for _, e := range status {
		if e.State != migrations.Applied {
			t.Fatalf("expected every entry Applied, got %+v", e)
		}
	}

	// No sample step overrides Down(): every rollback below is derived.
	reverted, err := runner.MigrateDown(ctx, 0, migrations.RunOptions{})
	if err != nil {
		t.Fatal(err)
	}
	if reverted != 10 {
		t.Fatalf("expected 10 reverted versions, got %d", reverted)
	}

	var objectCount int64
	if err := pool.QueryRowContext(ctx,
		"select count(*) from sqlite_master where type in ('table', 'view') "+
			"and name not like 'sqlite_%' and name <> 'schema_version'").Scan(&objectCount); err != nil {
		t.Fatal(err)
	}
	if objectCount != 0 {
		t.Fatalf("expected every sample object dropped, %d remain", objectCount)
	}

	// The same plan applies cleanly again: down really reached V0000.
	if applied, err = runner.Migrate(ctx, migrations.RunOptions{}); err != nil {
		t.Fatal(err)
	}
	if applied != 10 {
		t.Fatalf("expected 10 re-applied versions, got %d", applied)
	}
	roleNames, err = readRoleNames(ctx, pool)
	if err != nil {
		t.Fatal(err)
	}
	if len(roleNames) != 2 || roleNames[0] != "admin" || roleNames[1] != "user" {
		t.Fatalf("expected [admin user] after re-applying, got %v", roleNames)
	}
}

func readRoleNames(ctx context.Context, pool *sql.DB) ([]string, error) {
	rows, err := pool.QueryContext(ctx, "select role_name from roles order by role_name")
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var names []string
	for rows.Next() {
		var name string
		if err := rows.Scan(&name); err != nil {
			return nil, err
		}
		names = append(names, name)
	}
	return names, rows.Err()
}
